lua/entities/gmod_wire_expression2/base/ast.lua

```diff
@@ -0,0 +1,105 @@
+-- The E2 source code is parsed into a tree structure, known as an 'abstract
+-- syntax tree' or AST. This file provides utilities for operating on nodes of
+-- this tree generically.
+
+AddCSLuaFile()
+
+E2Lib.AST = {}
+
+local function genericChildVisitor(node, action)
+    for i = 3, #node do
+        local child = node[i]
+        if istable(child) and child.__instruction then
+            node[i] = action(child) or child
+        end
+    end
+end
+local childVisitors = {}
+function childVisitors.call(node, action)
+    local arguments = node[4]
+    for i, argument in ipairs(arguments) do
+        arguments[i] = action(argument) or argument
+    end
+end
+function childVisitors.methodcall(node, action)
+    local this, arguments = node[4], node[5]
+    node[4] = action(this) or this
+    for i, argument in ipairs(arguments) do
+        arguments[i] = action(argument) or argument
+    end
+end
+function childVisitors.stringcall(node, action)
+    local name, arguments = node[3], node[4]
+    node[3] = action(name) or name
+    for i, argument in ipairs(arguments) do
+        arguments[i] = action(argument) or argument
+    end
+end
+function childVisitors.kvtable(node, action)
+    local entries = node[3]
+    local additions, deletions = {}, {}
+    for key, value in pairs(entries) do
+        local newKey, newValue = action(key) or key, action(value) or value
+        if key ~= newKey then
+            additions[newKey] = newValue
+            table.insert(deletions, key)
```

Comment:
Assigning `nil` during a traversal won't cause any undefined behavior, so this deletion table isn't needed.

Reply:
Fixed.